Add capacity limiter to ConcurrentTaskRunner

jashbycu opened the pull request:
<!-- 
Thanks for opening a pull request to Prefect! We've got a few requests to help us review contributions:

- Make sure that your title neatly summarizes the proposed changes.
- Provide a short overview of the change and the value it adds.
- Share an example to help us understand the change in user experience.
- Confirm that you've done common tasks so we can give a timely review.

Happy engineering!
-->

Adds a CapacityLimiter to the ConcurrentTaskRunner to limit the number of concurrent threads. This is helpful for flows that would otherwise spawn hundreds of threads.

### Example

```
from prefect.task_runners import ConcurrentTaskRunner

@flow(task_runner=ConcurrentTaskRunner(max_workers=10))
def example_flow():
    # Flow that runs with at most 10 threads ...
```

### Checklist
<!-- These boxes may be checked after opening the pull request. -->

- [ ] This pull request references any related issue by including "closes `<link to issue>`"
	- If no issue exists and your change is not a small fix, please [create an issue](https://github.com/PrefectHQ/prefect/issues/new/choose) first.
- [ ] This pull request includes tests or only affects documentation.
- [x] This pull request includes a label categorizing the change e.g. `fix`, `feature`, `enhancement`
  <!--  If you do not have permission to add a label, a maintainer will add one for you and check this box. -->

madkinsz commented on src/prefect/task_runners.py:
```diff
@@ -323,6 +328,8 @@ async def _start(self, exit_stack: AsyncExitStack):
         """
         Start the process pool
         """
+        if self._max_workers is not None:
+            self._limiter = CapacityLimiter(self._max_workers)
```
Could we set this to an `asyncnullcontext` when `_max_workers` is null so that the code in `_run_and_store_result` doesn't need a branch?

madkinsz commented on src/prefect/task_runners.py:
```diff
@@ -65,6 +65,7 @@
 from uuid import UUID
 
 import anyio
+from anyio import CapacityLimiter
```
Can you use `anyio.CapacityLimiter` instead of importing the name? It's nice to have the clarity of where the implementation is.

madkinsz reviewed (changes requested):
Thanks for contributing! I've been meaning to add something like this. A couple comments on the implementation. We'll also need some sort of test before we can accept it.

madkinsz commented:
> Looking at test_task_runners, I'm not sure the best way to test this. Do we just need an additional test suite that yields (for example) `ConcurrentTaskRunner(max_workers=10)`?

That'd be a good start! It seems quite challenging to "prove" that it's being limited to some number of workers. Maybe would be useful to look at the AnyIO capacity limiter tests and see how they check it.

madkinsz commented on src/prefect/task_runners.py:
```diff
@@ -226,14 +227,14 @@ class ConcurrentTaskRunner(BaseTaskRunner):
         self._results: Dict[UUID, Any] = {}
         self._keys: Set[UUID] = set()
+        self._max_workers = max_workers
+        self._limiter = asyncnullcontext
```
```suggestion
        self._limiter = asyncnullcontext()
```

I'm also curious, is there a reason we need to instantiate the limiter in `_start` instead of `__init__`? Does it need the event loop?

madkinsz commented:
We might be able to set `max_workers=1` then assert that the task runner behaves sequentially instead of concurrently?

jashbycu commented:
Thanks Michael. I've pushed some test ideas, but I can't get the tests to run with `pytest tests/test_task_runners.py`. I get `pytest: error: unrecognized arguments: --cov=src/ --cov=tests/ --no-cov-on-fail --cov-report=term`. I don't see any detailed testing instructions in this repo. Can you help me?

One thing I'm not sure of is if `max_workers=1` will result in truly sequential task running, or if the ConcurrentTask Runner might run nearly-simultaneously-submitted tasks in an arbitrary order. So if the tests fail due to this, I have a workaround idea. But I have to be able to run the tests to see.

madkinsz commented on tests/test_task_runners.py:
```diff
@@ -32,3 +34,55 @@ class TestConcurrentTaskRunner(TaskRunnerStandardTestSuite):
+class TestConcurrentTaskRunnerSingleThreaded(TaskRunnerStandardTestSuite):
+    @pytest.fixture
+    def task_runner(self):
+        yield ConcurrentTaskRunner(max_workers=1)
```
You may need to / be able to change the the `concurrency_type` on this instance of the task runner from `TaskConcurrencyType.CONCURRENT` to `TaskConcurrencyType.SEQUENTIAL` so it goes through the sequential standard tests

madkinsz commented:
>  I've pushed some test ideas, but I can't get the tests to run with pytest tests/test_task_runners.py. I get pytest: error: unrecognized arguments: --cov=src/ --cov=tests/ --no-cov-on-fail --cov-report=term. I don't see any detailed testing instructions in this repo. Can you help me?

Sounds like you're missing some test requirements. Did you install with the `dev` extra? https://docs.prefect.io/contributing/overview/#setting-up-a-development-environment

> One thing I'm not sure of is if max_workers=1 will result in truly sequential task running, or if the ConcurrentTask Runner might run nearly-simultaneously-submitted tasks in an arbitrary order. So if the tests fail due to this, I have a workaround idea. But I have to be able to run the tests to see

Also not sure about what will happen there :) that seems like a possibility. If it's really a pain, we can explore a different test pattern.

jashbycu commented:
Ok I see that about the tests. I adjusted it to simply override the concurrency type, which hopefully will let it pass the tests. But I'm having trouble getting a dev env working on my computer, and unfortunately I have to focus on some other high priority work for me now. Short of tinkering with the tests, is there anything else I can do to help? I've added you as a collaborator to my fork.

madkinsz commented:
I've run the tests in CI and it looks like there are some issues. We don't have the bandwidth to prioritize this right now, but if it languishes we'll eventually get an engineer to get it across the line. Futzing with the tests is definitely the biggest part of this change :)

jashbycu commented:
Some more testing on my end shows that with more complex flows, this method clogs up and results in flows indefinitely stalling. I think it's because pending tasks count toward the thread limit. For my use case, I ended up using task tag concurrency limits, where each flow run I generate a unique tag to limit all tasks of that flow run, and then delete it when I'm done. I'm not sure what the Prefect solution should be, but maybe it needs to follow a similar pattern to that used by the task tags. In any case, I'm going to close this. Thank you!

madkinsz commented:
Thanks for investigating! It seems like we'd need to take the concurrency limit right before calling the user's function which is indeed trickier.
